ext/snowcrash/src/UriTemplateParser.cc

```diff
@@ -9,207 +9,251 @@
+                       pegtl::sor<
+                         pct_triplet,
+                         continuous_dots,
+                         invalid_char
+                       >
+                     > {};
+
+    template <typename Rule>
+    struct report_action : pegtl::nothing<Rule> {
+    };
+
+    template <>
+    struct report_action<invalid_char> {
+
+        template<typename Input>
+        static void apply(const Input& in, state& s)
+        {
+            s.position = in.position().byte;
+            s.length = in.string().length();
+
+            if (in.string() == "{") { // special case
+                s.message = std::string{"nested expression bracket '{'"};
+            } else {
+                s.message = std::string{"character '"} + in.string() + std::string{"'"};
             }
-            lastBracket = '{';
+
         }
-        if (uriTemplate[i] == '}') {
-            if (lastBracket == '}') {
-                result = true;
-                break;
-            }
-            lastBracket = '}';
+    };
+
+    template <>
+    struct report_action<continuous_dots> {
+
+        template<typename Input>
+        static void apply(const Input& in, state& s)
+        {
+            s.message = std::string{"continuous dots '"} + in.string() + std::string{"'"};
+            s.position = in.position().byte;
+            s.length = in.string().length();
         }
-    }
-    return result;
-}
+    };
 
-static bool PathContainsSquareBrackets(const URITemplate& uriTemplate)
-{
-    return (uriTemplate.find('[') != std::string::npos || uriTemplate.find(']') != std::string::npos);
-}
+    template <>
+    struct report_action<pct_triplet> {
 
-static Expressions GetUriTemplateExpressions(const URITemplate& uriTemplate)
-{
-    Expressions expressions;
-    size_t expressionStartPos = 0;
-    size_t expressionEndPos = 0;
-
-    while (expressionStartPos != std::string::npos && expressionEndPos != std::string::npos
-        && expressionStartPos < uriTemplate.length()) {
-        expressionStartPos = uriTemplate.find("{", expressionStartPos);
-        expressionEndPos = uriTemplate.find("}", expressionStartPos);
-        if (expressionStartPos != std::string::npos && expressionEndPos > expressionStartPos) {
-            expressions.push_back(
-                uriTemplate.substr(expressionStartPos + 1, (expressionEndPos - expressionStartPos) - 1));
+        template<typename Input>
+        static void apply(const Input& in, state& s)
+        {
+            s.message = std::string{"percent encoded character '"} + in.string() + std::string{"'"};
+            s.position = in.position().byte;
+            s.length = in.string().length();
         }
-        expressionStartPos++;
-    }
-    return expressions;
-}
+    };
 
-static ClassifiedExpression ClassifyExpression(const Expression& expression)
-{
+}
 
-    VariableExpression variableExpression(expression);
+struct VariableValidityChecker {
```

Comment:
Is there a specific reason we are using a callable function object instead of lambda-wrapped regular functions? Example:
```
void checkVariableValidity(...);
algorithm(begin(...), end(...), [](const auto& a){ checkVariableValidity(a, ...); })
```
Mostly it's possible to avoid `struct Foo{ ... operator()(...); }`, so that plain old functions are available; a lambda can be instantiated out of those building blocks in-line.

Reply:
It is called from `variant::visit` (resp. `mpark::visit`)
So it is common way how it is used. 

Or maybe I fully do not understand how do you mean your suggestion

Reply:
An example of what I have in mind, in the context of `variant`:
* delaration of specific C functions: https://github.com/apiaryio/drafter/blob/master/src/refract/JsonValue.cc#L83
  * notice the specific declarations for each variant version
  * notice the last declaration in that block (L96) that declaras a generalized function
* definition of the generalized function then uses the specific C functions and an inline lambda: https://github.com/apiaryio/drafter/blob/master/src/refract/JsonValue.cc#L302

There is no reason to define a named type `VariableValidityChecker` when we can have `checkVaraibleValidity` and wrap that in an anonymous type via a lambda where needed. The following benefits come to mind:

* readability
  * one less symbol
  * state of execution is more explicit if carried in function parameters, rather than in the body of a callable structure
* usability - a C function is more useful than a C++ object with special operators
* abstraction - a C function can be made into a callable object without cost, whereas reducing a callable object to a C function depends on compiler optimizations to avoid abstraction cost

Reply:
Oh, and to avoid confusion: The referenced example uses a named callable structure, but that structure's `operator()` just delegates the call completely to overloaded C functions. The reason for that is C++11 compatibility, back then there was no `auto` in lambdas (`[...](const auto&){...}`). So that's just a compatibility workaround.